sled-agent/src/bootstrap/agent.rs

```diff
@@ -236,91 +408,154 @@ impl Agent {
+                }
+                impl<'a> RestartMonitor<'a> {
+                    fn cancel(mut self) {
+                        self.run = false;
+                    }
+                }
+                impl<'a> Drop for RestartMonitor<'a> {
+                    fn drop(&mut self) {
+                        if self.run {
+                            *self.monitor = Some(HardwareMonitor::start(
+                                &self.log,
+                                self.hardware.take().unwrap(),
+                                self.services.take().unwrap(),
+                            ));
+                        }
+                    }
+                }
+                let restarter = RestartMonitor {
+                    run: true,
+                    log: self.log.clone(),
+                    hardware: Some(hardware),
+                    services: Some(services.clone()),
+                    monitor: hardware_monitor,
+                };
 
-            return Ok(SledAgentResponse { id: server.id() });
-        }
+                // Server does not exist, initialize it.
+                let server = SledServer::start(
+                    &self.sled_config,
+                    self.parent_log.clone(),
+                    request.clone(),
+                    services.clone(),
+                )
+                .await
+                .map_err(|e| {
+                    BootstrapError::SledError(format!(
+                        "Could not start sled agent server: {e}"
+                    ))
+                })?;
+                info!(&self.log, "Sled Agent loaded; recording configuration");
+
+                // Record this request so the sled agent can be automatically
+                // initialized on the next boot.
+                //
+                // danger handling: `serialized_request` contains our trust quorum
+                // share; we do not log it and only write it to the designated path.
+                let serialized_request = PersistentSledAgentRequest {
+                    request: Cow::Borrowed(request),
+                    trust_quorum_share: trust_quorum_share
+                        .clone()
+                        .map(Into::into),
+                }
+                .danger_serialize_as_toml()
+                .expect("Cannot serialize request");
 
-        if let Some(share) = trust_quorum_share.clone() {
-            self.establish_sled_quorum(share.clone()).await?;
-            *self.share.lock().await = Some(share);
-        }
+                let path = get_sled_agent_request_path();
+                tokio::fs::write(&path, &serialized_request).await.map_err(
+                    |err| BootstrapError::Io {
+                        message: format!(
+                            "Recording Sled Agent request to {path:?}"
+                        ),
+                        err,
+                    },
+                )?;
+
+                restarter.cancel();
```

Comment:
This is the point where we say: We aren't going to try to put the resources back; we're all-in on the sled agent starting.

Reply:
I'm going to just add this explicitly as a comment: 24a03b37e4d71bd8af502955e1ba111f8370aa04